### ISSUE TITLE
[a11y] Accessibility audit

### ISSUE THREAD

**Issue by aniamusial** [DivanteLtd/storefront-ui#960]
As StorefrontUI is a component library, we must remember that we can provide accessibility standards on component-level only. We can provide example usage on Pages, but most use-case scenarios (like keyboard trap) or setting tabindex must be handled by the user.

# MAIN GOALS 👩‍🏫
Key features to provide basic accessibility of components. Based on https://www.w3.org/TR/WCAG21/

## Keyboarding (2.1)
All functionalities should be available via keyboard
- make sure all interactive elements are focusable. As a user, I want to navigate through the page without using a mouse.
- make sure all interactive elements are clickable. As a user, I want to interact with the page without using a mouse.
- No keyboard trap. As a user, if I'm on filters view, I want to easily navigate through the rest of the page. If I'm on modal, I shouldn't be able to navigate to footer using the keyboard
- make sure user can use accelerator keys (space, enter) to interact with the page. 

## Semantics (1.3.1)

Good to read: https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML

When a screen reader, or any sort of assistive device scans a web page, it gets information about the Document Object Model (DOM), or the HTML structure of the page. No styles or JavaScript will be read by a screen reader. 

- make sure all element tags are used correctly for their purpose. buttons are `<button>`, not styled `<div>`
- make sure headers are used properly. There should be only one `<h1> `on each page
- links should be wrapped in` <a>`

## Descriptive (3.1)
- make sure all elements' texts are clear and understandable.
- avoid using symbolic elements like dash to describe things. f.e for the user with visual disabilities "5 to 10" is more descriptive than "5 - 10"
- make sure elements without text description has proper `aria-labels` and `alt text` set. As a user with vision disabilities, I want to access the elements on the page based on each element's descriptions. 
- use` aria-hidden="true"` for inline svg as the title of SVG files can sometimes be read by screen readers.
- make sure all links have the descriptive text: 
```
<!-- bad -->
<a href="/full-article">Read more</a>.

<!-- better -->
<a href="/full-article">Read more - Accessible Landmarks</a
```
- make sure all **form inputs** have descriptive error messages and info statements
```
<!-- bad -->
<span>Something went wrong</span>

<!-- better -->
<span>Password field must have at least 8 characters</span>
```
- make sure all form inputs have a label

- make sure the form is also validated with a general info message

# QUICK TIPS 🧠
- never remove focus outline, use v-focus directive instead and `tabindex`
- never set `tabindex` bigger than 0. It is a bad practice. Use` "0"` `"null"` or `"-1"` instead.
- do not use JS to imitate interactive elements event handling. If something is clickable, then it is a `button` / `link` etc.
- use `alt ` for non-decorative images, check [alt decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/)
- Accessible SVG patterns: https://www.deque.com/blog/creating-accessible-svgs/
- add `role="dialog"` for models to prevent keyboard trap
- if text is visible, do not use aria-label

# CURRENT ISSUES 🦙
- missing focus states on many components
- many icons act as a button, but are not `button` semantically.
- many `divs` act as button

# ACTION POINTS 🚀
## Implement` v-focus` directive on interactive elements.
See [SfColor](https://storybook.storefrontui.io/?path=/story/atoms-color--common) for reference. 
Remember to set proper `tabindex`.

Related issue: 

Components to include: 

ATOMS:
- [x] SfArrow @justyna-13 
- [x] SfButton @justyna-13 
- [x] SfBreadcrumbs - links 
- [x] SfBullets 
- [x] SfColor
- [ ] SfCheckbox - checkbox @justyna-13 @pspaczek 
- [x] SfChevron @justyna-13 
- [x] SfCircleIcon @justyna-13 
- [x] SfIcon
- [x] SfInput @justyna-13 
- [x] SfQuantitySelector @justyna-13 

MOLECULES:
- [x] SfAddToCart - input and button
- [x] SfBanner - button @justyna-13 
- [x] SfBar - back button and close button @justyna-13 
- [x] SfCallToAction - button @justyna-13 
- [ ] SfFilter - checkbox @justyna-13 
- [x] SfGallery - thumbs 
- [x] SfModal - close button @justyna-13 
- [x] SfNotification - close button and text button @justyna-13 
- [x] SfPagination - navigation arrows and steps @martaradziszewska 
- [x] SfRadio - radio button checkmark @martaradziszewska
- [x] SfScrollable - text buttons @martaradziszewska 
- [x] SfSearch - input @martaradziszewska 
- [x] SfSelect - input and dropdown items @martaradziszewska 
- [x] SfSteps - steps @martaradziszewska 

ORGANISMS:
- [x] SfAccordion - accordion items and chevrons @martaradziszewska 
- [x] SfBannerGrid - button and each banner (as it is clickable) on mobile @justyna-13 
- [x] SfBottomNavigation - bottom navigation items and addToCart button @justyna-13 
- [x] SfCarousel - navigation arrows and carousel items @justyna-13 
- [x] SfCollectedProduct - collected product as a whole, quantity selector, close button on mobile @justyna-13 
- [x] SfContentPages - tabs and menu items @justyna-13 
- [x] SfFooter - footer column title and footer column items, social media icons @justyna-13 
- [x] SfGroupedProduct - quantity selector @justyna-13 
- [x] SfHeader - header items, logo, search, icons @justyna-13 
- [x] SfHero - navigation arrows, button, hero image on mobile @justyna-13 
- [x] SfList - list items @justyna-13 
- [x] SfMegaMenu - mega menu items, mega menu title, mega menu tabs on mobile @justyna-13 
- [x] SfProductCard - product card as whole, circle button, reviews, wishlist icon @martaradziszewska
- [x] SfProductCardHorizontal - quantity selector, productcard as whole, review, text buttons,  add to cart button, favorite icon @martaradziszewska 
- [x] SfSidebar - close button @martaradziszewska 
- [x] SfStoreLocator - map controls, store items, phone number and email.
- [x] SfTable - table rows @justyna-13 
- [x] SfTabs - tabs
- [x] SfTopBar - text buttons, location icon

How to test keyboard navigation: https://a11yproject.com/posts/navigate-using-just-your-keyboard/

## Fix semantics of components

Related issue: 

- [x] SfSteps - steps should be a `<button>` instead of `<div>`
- [x] SfStoreLocator - phone number and email should be a link `<a>`
- [x] SfBreadcrumbs - list elements should be links `<a>` @justyna-13  @pspaczek 
- [x] SfTabs - tabs should be a `<button>` instead of `<div>`
- [x] SfBullets - list elements should be buttons `<button>` with `aria` @justyna-13 
- [x] SfCheckbox - missing `type="checkbox"` for `<input>`
- [x] SfLoader - svg misses` role="img"` and `<title>` @martaradziszewska 
- [ ] SfPagination - `<buttons> `for navigation must be an` <a>` ! @martaradziszewska @pspaczek 

@pspaczek please check SfPagination and SfBreadcrumbs IMO we can't change it to <a> elements

## Elements must be descriptive

Related issue: 

- [ ] SfButton should have `ariaPressed` prop to set `aria-pressed` in case someone would like to define the button as a toggle. @martaradziszewska  / @aniamusial do we need to do that? attributes are inheritable thanks to `v-bind="$attrs"`
- [ ] SfIcon should have`aria-hidden` prop - in case the inside svg is purely decorative @justyna-13 
- [x] SfGallery - gallery thumbs images misses `alt` text @justyna-13 
- [x] SfTopBar stories - `alt` text for flag icon should be more descriptive @justyna-13 
- [x] Cart - empty cart image misses `alt` text @justyna-13 
- [x] SfColor - `SfIcon` inside` SfBadge` should have `aria-hidden="true"` @martaradziszewska 
- [ ] SfIcon - we should discuss whether to add `ariaPressed` prop for `SfIcon` @martaradziszewska 
- [x] SfAddToCart - misses `aria-labelledby `
- [x] SfIcon in SfAlert misses `aria-hidden` @martaradziszewska
- [x] SfCharacteristic misses `aria-hidden` @martaradziszewska

## SfIcon needs to act as a button in some cases

Related issue: 

SfIcon should be a button with `aria` for many cases, when it is interactive. We should think of solution - a prop maybe?

**Comment by Qrzy** [comment]
Hello @aniamusial, really impressive piece of work has been done here!

I've got just one comment that could possibly bring focus to some missed point.
Regarding the `Elements must be descriptive` section, it's been reported that:
> button/action names have the same name eg. `<button aria-label="Add to wishlist">` - this will result in confusion as to which product they are adding to the wishlist

Do you think the product card component should use product title in labels for actions like adding to wishlist?

**Comment by justyna-13** [comment]
SfIcon - if it's clickable shouldn't have rele="button". Instead we should use button tag
- [ ] SfBar @justyna-13
- [ ] SfModal @justyna-13
- [ ] SfCollectedProduct @justyna-13
- [ ] SfHeader @justyna-13
SfCircleicon is a button, so that shouldn't have role="button" 
- [ ] SfProductCard @justyna-13
